From cb4533bbd853a4448bb37b1ee3b38927690ba8e5 Mon Sep 17 00:00:00 2001
From: Wei Qin <qinwei61@huawei.com>
Date: Tue, 3 Mar 2026 09:37:32 +0800
Subject: [PATCH 2/2] ub: udma: Support user ctl function.

driver inclusion
category: bugfix
bugzilla: https://atomgit.com/openeuler/umdk/issues/160
CVE: NA

-----------------------------------------------------------

This patch support user ctl function.

Fixes: a7ea7e683ec6 ("udma: support alloc and free tid")

Signed-off-by: Wei Qin <qinwei61@huawei.com>
---
 src/urma/hw/udma/kernel_headers/udma_abi.h |  11 +
 src/urma/hw/udma/udma_u_buf.c              |  13 +-
 src/urma/hw/udma/udma_u_common.h           |   9 +-
 src/urma/hw/udma/udma_u_ctl.c              |   9 +-
 src/urma/hw/udma/udma_u_jetty.c            | 538 +++++++++++++++++++--
 src/urma/hw/udma/udma_u_jetty.h            |  36 ++
 src/urma/hw/udma/udma_u_jfc.c              | 283 ++++++++++-
 src/urma/hw/udma/udma_u_jfc.h              |   6 +
 src/urma/hw/udma/udma_u_jfr.c              | 375 +++++++++++++-
 src/urma/hw/udma/udma_u_jfr.h              |  17 +
 src/urma/hw/udma/udma_u_jfs.c              | 314 +++++++++++-
 src/urma/hw/udma/udma_u_jfs.h              |  10 +
 src/urma/hw/udma/udma_u_ops.c              |  79 ++-
 13 files changed, 1630 insertions(+), 70 deletions(-)

diff --git a/src/urma/hw/udma/kernel_headers/udma_abi.h b/src/urma/hw/udma/kernel_headers/udma_abi.h
index 5d2cad0..25ac18d 100644
--- a/src/urma/hw/udma/kernel_headers/udma_abi.h
+++ b/src/urma/hw/udma/kernel_headers/udma_abi.h
@@ -50,6 +50,7 @@ enum udma_jetty_type {
 	UDMA_CCU_JETTY_TYPE,
 	UDMA_NORMAL_JETTY_TYPE,
 	UDMA_URMA_NORMAL_JETTY_TYPE,
+	UDMA_URMA_EX_JETTY_TYPE,
 	UDMA_JETTY_TYPE_MAX
 };
 
@@ -82,6 +83,10 @@ struct udma_create_jetty_ucmd {
 	__u32 rsv1;
 };
 
+struct udma_create_jetty_resp {
+	__aligned_u64 buf_addr;
+};
+
 struct udma_create_jfc_ucmd {
 	__aligned_u64 buf_addr;
 	__u32 buf_len;
@@ -97,6 +102,7 @@ struct udma_create_ctx_resp {
 	__u32 dwqe_enable : 1;
 	__u32 reduce_enable : 1;
 	__u32 dump_aux_info : 1;
+	__u32 sq_reserved : 1;
 	__u32 hugepage_enable : 1;
 	__u32 sva_sep_mode_en : 1;
 	__u32 rsv : 19;
@@ -104,7 +110,11 @@ struct udma_create_ctx_resp {
 	__u32 chip_id;
 	__u32 die_id;
 	__u32 jfr_sge;
+	__u32 ccu_jetty_start_id;
+	__u32 ccu_jetty_max_cnt;
 	__u32 rsv1;
+	__aligned_u64 sq_reserved_va;
+	__aligned_u64 sq_reserved_len;
 };
 
 struct udma_create_jfr_resp {
@@ -116,6 +126,7 @@ enum db_mmap_type {
 	UDMA_MMAP_HUGEPAGE,
 	UDMA_MMAP_JFC_PAGE,
 	UDMA_MMAP_JETTY_DSQE,
+	UDMA_MMAP_RESERVED_SQ,
 	UDMA_MMAP_KERNEL_BUF,
 };
 
diff --git a/src/urma/hw/udma/udma_u_buf.c b/src/urma/hw/udma/udma_u_buf.c
index e2d1097..b23f806 100644
--- a/src/urma/hw/udma/udma_u_buf.c
+++ b/src/urma/hw/udma/udma_u_buf.c
@@ -54,7 +54,13 @@ bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt
 		}
 	}
 
-	if (q->ctx->hugepage_enable) {
+	if (q->cstm)
+		goto alloc_queue_buf_finish;
+
+	if (q->sq_reserved)
+		return true;
+
+	if (!q->cstm && q->ctx->hugepage_enable) {
 		q->hugepage = udma_u_alloc_hugepage(q->ctx, q->qbuf_size);
 		if (q->hugepage) {
 			q->qbuf = q->hugepage->va_start;
@@ -62,10 +68,11 @@ bool udma_u_alloc_queue_buf(struct udma_u_jetty_queue *q, uint32_t max_entry_cnt
 			UDMA_LOG_WARN("failed to alloc hugepage buf, switch to alloc normal buf.\n");
 			q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 		}
-	} else {
+	} else if (!q->cstm) {
 		q->qbuf = udma_u_alloc_kernel_buf(q->ctx, q->qbuf_size);
 	}
 
+alloc_queue_buf_finish:
 	if (!q->qbuf) {
 		UDMA_LOG_ERR("failed to alloc queue buffer.\n");
 		if (wrid_en) {
@@ -87,7 +94,7 @@ void udma_u_free_queue_buf(struct udma_u_jetty_queue *q)
 		q->wrid = NULL;
 	}
 
-	if (q->cstm)
+	if (q->cstm || q->sq_reserved)
 		return;
 
 	if (q->qbuf != NULL) {
diff --git a/src/urma/hw/udma/udma_u_common.h b/src/urma/hw/udma/udma_u_common.h
index bb9f855..7a8121d 100644
--- a/src/urma/hw/udma/udma_u_common.h
+++ b/src/urma/hw/udma/udma_u_common.h
@@ -96,7 +96,10 @@ struct udma_u_context {
 	uint32_t		die_id;
 	bool			dump_aux_info;
 	uint32_t		jfr_sge;
+	bool			sq_reserved;
 	bool			sva_sep_mode_en;
+	uint32_t		ccu_jetty_start_id;
+	uint32_t		ccu_jetty_max_cnt;
 	bool			hugepage_enable;
 	pthread_mutex_t		hugepage_lock;
 	struct udma_u_hugepage_priv *hugepage_list;
@@ -141,7 +144,11 @@ struct udma_u_jetty_queue {
 	uint32_t old_entry_idx;
 	bool lock_free;
 	bool cstm; /* sq ctrl flag */
+	bool sq_reserved;
 	struct udma_u_hugepage *hugepage;
+	uint8_t db_status;
+	bool need_ring_db;
+	bool pi_type;
 };
 
 struct udma_wqe_sge {
@@ -176,7 +183,6 @@ struct udma_u_jfr {
 struct udma_u_jfs {
 	urma_jfs_t base;
 	struct udma_u_jetty_queue sq;
-	bool pi_type;
 	uint32_t jfs_type;
 };
 
@@ -186,7 +192,6 @@ struct udma_u_jetty {
 	struct udma_u_jfr *jfr;
 	struct udma_u_jetty_grp *jetty_grp;
 	uint32_t jetty_type;
-	bool pi_type;
 };
 
 struct udma_u_jfc {
diff --git a/src/urma/hw/udma/udma_u_ctl.c b/src/urma/hw/udma/udma_u_ctl.c
index d735684..8f623df 100644
--- a/src/urma/hw/udma/udma_u_ctl.c
+++ b/src/urma/hw/udma/udma_u_ctl.c
@@ -86,6 +86,9 @@ static int udma_u_init_jfs_queue_buf(struct udma_u_jetty_queue *q,
 		return ENOMEM;
 	}
 
+	if (!cstm && q->sq_reserved)
+		return 0;
+
 	if (cstm) {
 		q->qbuf = cfg_ex->cstm_cfg.sq.buff;
 	} else if (q->ctx->hugepage_enable) {
@@ -379,13 +382,14 @@ static urma_jfs_t *udma_u_create_jfs_ex(urma_context_t *ctx,
 	}
 
 	jfs->sq.ctx = udma_ctx;
+	jfs->sq.sq_reserved = udma_ctx->sq_reserved;
 	if (udma_u_create_sq_ex(&jfs->sq, cfg_ex)) {
 		UDMA_LOG_ERR("failed to create sq.\n");
 		goto err_create_sq;
 	}
 
 	jfs->sq.cstm = cfg_ex->cstm_cfg.flag.bs.sq_cstm;
-	jfs->pi_type = cfg_ex->pi_type;
+	jfs->sq.pi_type = cfg_ex->pi_type;
 	jfs->jfs_type = (uint32_t)cfg_ex->jetty_type;
 	jfs->sq.db.id = cfg_ex->id;
 	cfg_ex->base_cfg.depth = 1;
@@ -515,7 +519,7 @@ static void udma_u_init_jetty_param_ex(struct udma_u_jetty *jetty,
 	jetty->jfr = to_udma_u_jfr(cfg_ex->base_cfg.shared.jfr);
 	jetty->sq.cstm = cfg_ex->jfs_cstm.flag.bs.sq_cstm != 0 ? true : false;
 	jetty->jetty_type = (uint32_t)cfg_ex->jetty_type;
-	jetty->pi_type = cfg_ex->pi_type;
+	jetty->sq.pi_type = cfg_ex->pi_type;
 	cfg_ex->base_cfg.jfs_cfg.depth = 1;
 }
 
@@ -540,6 +544,7 @@ static urma_jetty_t *udma_u_create_jetty_ex(urma_context_t *ctx, struct udma_u_j
 
 	udma_u_init_jfs_cfg_ex(cfg_ex, &jfs_cfg);
 	jetty->sq.ctx = udma_ctx;
+	jetty->sq.sq_reserved = udma_ctx->sq_reserved;
 	ret = udma_u_create_sq_ex(&jetty->sq, &jfs_cfg);
 	if (ret != 0) {
 		UDMA_LOG_ERR("jetty create sq failed, ret = %d.\n", ret);
diff --git a/src/urma/hw/udma/udma_u_jetty.c b/src/urma/hw/udma/udma_u_jetty.c
index 67e8a5f..e5a88e2 100644
--- a/src/urma/hw/udma/udma_u_jetty.c
+++ b/src/urma/hw/udma/udma_u_jetty.c
@@ -18,9 +18,9 @@
 #include "udma_u_db.h"
 #include "udma_u_jetty.h"
 
-int exec_jetty_create_cmd(urma_context_t *ctx, struct udma_u_jetty *jetty,
-			  urma_jetty_cfg_t *cfg)
+static int exec_jetty_active_cmd(struct udma_u_jetty *jetty, urma_jetty_cfg_t *cfg)
 {
+	struct udma_create_jetty_resp resp = {};
 	struct udma_create_jetty_ucmd cmd = {};
 	urma_cmd_udrv_priv_t udata = {};
 	int ret;
@@ -36,17 +36,64 @@ int exec_jetty_create_cmd(urma_context_t *ctx, struct udma_u_jetty *jetty,
 	cmd.db_addr = (uintptr_t)jetty->sq.db.addr;
 	cmd.jetty_addr = (uintptr_t)&jetty->sq;
 	cmd.sqe_bb_cnt = jetty->sq.sqe_bb_cnt;
-	cmd.pi_type = jetty->pi_type;
+	cmd.pi_type = jetty->sq.pi_type;
 	cmd.jetty_type = jetty->jetty_type;
 	cmd.non_pin = jetty->sq.cstm;
 	cmd.is_hugepage = jetty->sq.hugepage != NULL;
 
-	udma_u_set_udata(&udata, &cmd, (uint32_t)sizeof(cmd), NULL, 0);
+	udma_u_set_udata(&udata, &cmd, (uint32_t)sizeof(cmd), &resp, (uint32_t)sizeof(resp));
+	ret = urma_cmd_active_jetty(&jetty->base, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("urma active jetty failed.\n");
+		return ret;
+	}
+
+	ret = udma_u_set_sq_by_resp(&jetty->sq, &resp);
+	if (ret != 0) {
+		(void)urma_cmd_deactive_jetty(&jetty->base, NULL);
+		return ret;
+	}
+
+	jetty->sq.idx = jetty->base.jetty_id.id;
+	jetty->sq.db.id = jetty->base.jetty_id.id;
+
+	return 0;
+}
+
+int exec_jetty_create_cmd(urma_context_t *ctx, struct udma_u_jetty *jetty,
+				 urma_jetty_cfg_t *cfg)
+{
+	struct udma_create_jetty_resp resp = {};
+	struct udma_create_jetty_ucmd cmd = {};
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	if (cfg->jfs_cfg.priority >= UDMA_MAX_PRIORITY) {
+		UDMA_LOG_ERR("user mode jetty priority is out of range, priority is %u.\n",
+			     cfg->jfs_cfg.priority);
+		return EINVAL;
+	}
+
+	cmd.buf_addr = (uintptr_t)jetty->sq.qbuf;
+	cmd.buf_len = jetty->sq.qbuf_size;
+	cmd.db_addr = (uintptr_t)jetty->sq.db.addr;
+	cmd.jetty_addr = (uintptr_t)&jetty->sq;
+	cmd.sqe_bb_cnt = jetty->sq.sqe_bb_cnt;
+	cmd.pi_type = jetty->sq.pi_type;
+	cmd.jetty_type = jetty->jetty_type;
+	cmd.non_pin = jetty->sq.cstm;
+	cmd.is_hugepage = jetty->sq.hugepage != NULL;
+	udma_u_set_udata(&udata, &cmd, (uint32_t)sizeof(cmd), &resp, (uint32_t)sizeof(resp));
 	ret = urma_cmd_create_jetty(ctx, &jetty->base, cfg, &udata);
 	if (ret) {
 		UDMA_LOG_ERR("urma create jetty failed.\n");
 		return ret;
 	}
+	ret = udma_u_set_sq_by_resp(&jetty->sq, &resp);
+	if (ret != 0) {
+		(void)urma_cmd_delete_jetty(&jetty->base);
+		return ret;
+	}
 
 	jetty->sq.idx = jetty->base.jetty_id.id;
 	jetty->sq.db.id = jetty->base.jetty_id.id;
@@ -79,7 +126,7 @@ int add_jetty_to_grp(struct udma_u_jetty *jetty, urma_jetty_cfg_t *cfg)
 		return 0;
 
 	if (jetty->sq.trans_mode != URMA_TM_RM) {
-		UDMA_LOG_ERR("Jetty must be RM model, if assigned grp.\n");
+		UDMA_LOG_ERR("jetty must be RM model, if assigned grp.\n");
 		return EINVAL;
 	}
 
@@ -87,7 +134,7 @@ int add_jetty_to_grp(struct udma_u_jetty *jetty, urma_jetty_cfg_t *cfg)
 
 	(void)pthread_spin_lock(&udma_jetty_grp->lock);
 	if (udma_jetty_grp->jetty_cnt >= MAX_JETTY_IN_GRP) {
-		UDMA_LOG_ERR("Jetty Group is aleady full.\n");
+		UDMA_LOG_ERR("jetty Group is already full.\n");
 		ret = EINVAL;
 		goto out;
 	}
@@ -145,6 +192,37 @@ urma_status_t insert_jetty_node(struct udma_u_context *udma_ctx,
 	return URMA_SUCCESS;
 }
 
+static int udma_u_active_jetty_prepare(urma_jetty_t *jetty, urma_jetty_cfg_t *cfg)
+{
+	struct udma_u_jetty *udma_jetty = to_udma_u_jetty(jetty);
+	int ret;
+
+	ret = init_jetty_trans_mode(udma_jetty, cfg);
+	if (ret) {
+		UDMA_LOG_ERR("init jetty tm failed when active jetty, ret = %d.\n", ret);
+		return ret;
+	}
+
+	ret = add_jetty_to_grp(udma_jetty, cfg);
+	if (ret) {
+		UDMA_LOG_ERR("active jetty add jetty to grp failed, ret = %d.\n", ret);
+		return ret;
+	}
+
+	udma_jetty->sq.sq_reserved = udma_jetty->sq.ctx->sq_reserved;
+	udma_jetty->sq.db.id = jetty->jetty_cfg.id;
+	ret = udma_u_create_sq(&udma_jetty->sq, &cfg->jfs_cfg);
+	if (ret) {
+		UDMA_LOG_ERR("active jetty create sq failed, ret = %d.\n", ret);
+		remove_jetty_from_grp(udma_jetty);
+		return ret;
+	}
+
+	udma_jetty->jfr = to_udma_u_jfr(cfg->shared.jfr);
+
+	return 0;
+}
+
 urma_jetty_t *udma_u_create_jetty(urma_context_t *ctx, urma_jetty_cfg_t *cfg)
 {
 	struct udma_u_context *udma_ctx = to_udma_u_ctx(ctx);
@@ -157,26 +235,12 @@ urma_jetty_t *udma_u_create_jetty(urma_context_t *ctx, urma_jetty_cfg_t *cfg)
 		return NULL;
 	}
 
-	ret = init_jetty_trans_mode(jetty, cfg);
-	if (ret) {
-		UDMA_LOG_ERR("init jetty transmode failed.\n");
-		goto err_add_to_grp;
-	}
-
-	ret = add_jetty_to_grp(jetty, cfg);
-	if (ret) {
-		UDMA_LOG_ERR("add Jetty to grp failed.\n");
-		goto err_add_to_grp;
-	}
-
 	jetty->sq.ctx = udma_ctx;
-	ret = udma_u_create_sq(&jetty->sq, &cfg->jfs_cfg);
-	if (ret) {
-		UDMA_LOG_ERR("Jetty create sq failed.\n");
-		goto err_create_sq;
-	}
 
-	jetty->jfr = to_udma_u_jfr(cfg->shared.jfr);
+	ret = udma_u_active_jetty_prepare(&jetty->base, cfg);
+	if (ret)
+		goto err_active_jetty_prepare;
+
 	jetty->jetty_type = UDMA_URMA_NORMAL_JETTY_TYPE;
 	if (exec_jetty_create_cmd(ctx, jetty, cfg)) {
 		UDMA_LOG_ERR("failed to create jetty.\n");
@@ -199,9 +263,8 @@ err_alloc_db:
 	(void)urma_cmd_delete_jetty(&jetty->base);
 err_jetty_create_cmd:
 	udma_u_delete_sq(&jetty->sq);
-err_create_sq:
 	remove_jetty_from_grp(jetty);
-err_add_to_grp:
+err_active_jetty_prepare:
 	free(jetty);
 
 	return NULL;
@@ -220,11 +283,15 @@ void udma_u_jetty_table_remove(struct udma_u_context *udma_ctx,
 		return;
 	}
 
+	if (!udma_ctx->jetty_table[table_id].jetty_array[jetty->sq.idx & mask]) {
+		(void)pthread_rwlock_unlock(&udma_ctx->jetty_table_lock);
+		return;
+	}
+	udma_ctx->jetty_table[table_id].jetty_array[jetty->sq.idx & mask] = NULL;
+
 	if (!--udma_ctx->jetty_table[table_id].refcnt) {
 		free(udma_ctx->jetty_table[table_id].jetty_array);
 		udma_ctx->jetty_table[table_id].jetty_array = NULL;
-	} else {
-		udma_ctx->jetty_table[table_id].jetty_array[jetty->sq.idx & mask] = NULL;
 	}
 	(void)pthread_rwlock_unlock(&udma_ctx->jetty_table_lock);
 }
@@ -246,7 +313,7 @@ static urma_status_t udma_u_delete_jetty_prepare(urma_jetty_t *jetty)
 	return URMA_SUCCESS;
 }
 
-static void udma_u_free_jetty(urma_jetty_t *jetty)
+static void udma_u_free_jetty_prepare(urma_jetty_t* jetty)
 {
 	struct udma_u_context *udma_ctx = to_udma_u_ctx(jetty->urma_ctx);
 	struct udma_u_jetty *udma_jetty = to_udma_u_jetty(jetty);
@@ -270,7 +337,12 @@ static void udma_u_free_jetty(urma_jetty_t *jetty)
 	udma_u_free_db(jetty->urma_ctx, &udma_jetty->sq.db);
 	udma_u_delete_sq(&udma_jetty->sq);
 	remove_jetty_from_grp(udma_jetty);
-	free(udma_jetty);
+}
+
+static void udma_u_free_jetty_detail(urma_jetty_t *jetty)
+{
+	udma_u_free_jetty_prepare(jetty);
+	free(jetty);
 }
 
 urma_status_t udma_u_delete_jetty(urma_jetty_t *jetty)
@@ -289,7 +361,7 @@ urma_status_t udma_u_delete_jetty(urma_jetty_t *jetty)
 		goto delete_err;
 	}
 
-	udma_u_free_jetty(jetty);
+	udma_u_free_jetty_detail(jetty);
 
 	return URMA_SUCCESS;
 
@@ -331,7 +403,7 @@ urma_status_t udma_u_delete_jetty_batch(urma_jetty_t **jetty, int jetty_cnt, urm
 	}
 
 	for (i = 0; i < jetty_cnt; i++)
-		udma_u_free_jetty(jetty[i]);
+		udma_u_free_jetty_detail(jetty[i]);
 
 	return 0;
 
@@ -664,3 +736,405 @@ urma_status_t udma_u_bind_jetty_ex(urma_jetty_t *jetty,
 
 	return URMA_SUCCESS;
 }
+
+urma_status_t udma_u_alloc_jetty(urma_context_t *ctx, urma_jetty_cfg_t *cfg,
+				urma_jetty_t **jetty)
+{
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(ctx);
+	struct udma_u_jetty *udma_jetty;
+	urma_cmd_udrv_priv_t udata;
+	int ret;
+
+	udma_jetty = (struct udma_u_jetty *)calloc(1, sizeof(struct udma_u_jetty));
+	if (udma_jetty == NULL) {
+		UDMA_LOG_ERR("udma jetty allocation failed.\n");
+		return URMA_ENOMEM;
+	}
+
+	ret = urma_cmd_alloc_jetty(ctx, cfg, &udma_jetty->base, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("alloc jetty failed, ret = %d.\n", ret);
+		free(udma_jetty);
+		return URMA_FAIL;
+	}
+
+	*jetty = &udma_jetty->base;
+	udma_jetty->sq.ctx = udma_ctx;
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_free_jetty(urma_jetty_t *jetty)
+{
+	urma_cmd_udrv_priv_t udata;
+	int ret;
+
+	ret = urma_cmd_free_jetty(jetty, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("free jetty failed, ret = %d.\n", ret);
+		return URMA_FAIL;
+	}
+
+	free(jetty);
+
+	return URMA_SUCCESS;
+}
+
+static uint64_t udma_jetty_opt[] = {
+	URMA_JFS_DEPTH,
+	URMA_JFS_FLAG,
+	URMA_JFS_TRANS_MODE,
+	URMA_JFS_PRIORITY,
+	URMA_JFS_MAX_SGE,
+	URMA_JFS_MAX_RSGE,
+	URMA_JFS_MAX_INLINE_DATA,
+	URMA_JFS_RNR_RETRY,
+	URMA_JFS_ERR_TIMEOUT,
+	URMA_JFS_BIND_JFC,
+	URMA_JFS_USER_CTX,
+	URMA_JFS_SQE_BASE_ADDR,
+	URMA_JFS_ID,
+	URMA_JFS_DB_ADDR,
+	URMA_JFS_DB_STATUS,
+	URMA_JFS_PI,
+	URMA_JFS_PI_TYPE,
+	URMA_JFS_CI,
+	URMA_JETTY_ID,
+	URMA_JETTY_FLAG,
+	URMA_JETTY_BIND_JFR,
+	URMA_JETTY_BIND_RX_JFC,
+	URMA_JETTY_BIND_JTG,
+	URMA_JETTY_USER_CTX,
+};
+
+static struct udma_u_jetty_opt_info opt_info[] = {
+	{sizeof(uint32_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint32_t), PERM_R | PERM_W, JETTY_IGNORE, JFS_MODE | JETTY_MODE},
+	{sizeof(uint32_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint8_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint8_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint8_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint32_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint8_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint8_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint64_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint64_t), PERM_R | PERM_W, JFS_IGNORE | JETTY_IGNORE, JFS_MODE | JETTY_MODE},
+	{sizeof(uint64_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint32_t), PERM_R | PERM_W, 0, JFS_MODE},
+	{sizeof(uint64_t), PERM_R, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint8_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint16_t), PERM_R, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint16_t), PERM_R | PERM_W, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint16_t), PERM_R, 0, JFS_MODE | JETTY_MODE},
+	{sizeof(uint32_t), PERM_R | PERM_W, 0, JETTY_MODE},
+	{sizeof(uint32_t), PERM_R | PERM_W, 0, JETTY_MODE},
+	{sizeof(uint64_t), PERM_R | PERM_W, 0, JETTY_MODE},
+	{sizeof(uint64_t), PERM_R | PERM_W, 0, JETTY_MODE},
+	{sizeof(uint64_t), PERM_R | PERM_W, 0, JETTY_MODE},
+	{sizeof(uint64_t), PERM_R | PERM_W, JETTY_IGNORE, JETTY_MODE},
+};
+
+urma_status_t udma_u_verify_jetty_opt(uint64_t opt, void *buf, uint32_t len,
+	enum udma_u_set_get_jetty_opt_mode jetty_mode,
+	enum udma_u_set_get_jetty_opt_perm jetty_perm)
+{
+	uint32_t opt_index = ARRAY_SIZE(udma_jetty_opt);
+	uint32_t i;
+
+	if (buf == NULL) {
+		UDMA_LOG_ERR("jetty/jfs opt buf is null\n");
+		return URMA_EINVAL;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(udma_jetty_opt); i++) {
+		if (udma_jetty_opt[i] == opt) {
+			opt_index = i;
+			break;
+		}
+	}
+
+	if (opt_index >= ARRAY_SIZE(opt_info) || (opt_info[opt_index].mode & jetty_mode) == 0) {
+		UDMA_LOG_ERR("opt %lu mode %u is invalid\n", opt, jetty_mode);
+		return URMA_EINVAL;
+	}
+
+	if (len != opt_info[opt_index].buf_len) {
+		UDMA_LOG_ERR("opt %lu len %u error, should be %u\n",
+			     opt, len, opt_info[opt_index].buf_len);
+		return URMA_EINVAL;
+	}
+
+	if ((opt_info[opt_index].perm & jetty_perm) == 0) {
+		UDMA_LOG_ERR("opt %lu not allow read\n", opt);
+		return URMA_EINVAL;
+	}
+
+	/* The EEXIST return value indicates that there is no need to enter kernel mode. */
+	if (opt_info[opt_index].ignore_attr & jetty_mode) {
+		UDMA_LOG_ERR("opt %lu no need to trap kernel\n", opt);
+		return URMA_EEXIST;
+	}
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_set_jetty_field(struct udma_u_jetty_queue *sq, uint64_t opt, void *buf)
+{
+	uint8_t in_data;
+
+	switch (opt) {
+	case URMA_JFS_SQE_BASE_ADDR:
+		sq->qbuf = (void *)(uintptr_t)(*((uint64_t *)buf));
+		if (!sq->qbuf) {
+			UDMA_LOG_ERR("set sqe base addr is null.\n");
+			return URMA_EINVAL;
+		}
+		sq->cstm = true;
+		break;
+	case URMA_JFS_DB_STATUS:
+		in_data = *((uint8_t *)buf);
+		if (in_data > 1) {
+			UDMA_LOG_ERR("db status %hhu is invalid\n", in_data);
+			return URMA_EINVAL;
+		}
+		sq->db_status = in_data;
+		if (sq->db_status == 0 && sq->need_ring_db) {
+			if (!sq->lock_free)
+				(void)pthread_spin_lock(&sq->lock);
+
+			sq->need_ring_db = false;
+			udma_update_sq_db(sq);
+
+			if (!sq->lock_free)
+				(void)pthread_spin_unlock(&sq->lock);
+		}
+		break;
+	case URMA_JFS_PI_TYPE:
+		in_data = *((uint8_t *)buf);
+		if (in_data > 1) {
+			UDMA_LOG_ERR("db status %hhu is invalid\n", in_data);
+			return URMA_EINVAL;
+		}
+		sq->pi_type = in_data;
+		break;
+	default:
+		break;
+	}
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_set_jetty_opt(urma_jetty_t *jetty, uint64_t opt, void *buf, uint32_t len)
+{
+	struct udma_u_jetty *udma_jetty = to_udma_u_jetty(jetty);
+	urma_cmd_udrv_priv_t udata;
+	urma_status_t verify_ret;
+	urma_status_t urma_ret;
+	int ret;
+
+	verify_ret = udma_u_verify_jetty_opt(opt, buf, len, JETTY_MODE, PERM_W);
+	if (verify_ret && verify_ret != URMA_EEXIST)
+		return verify_ret;
+
+	urma_ret = udma_u_set_jetty_field(&udma_jetty->sq, opt, buf);
+	if (urma_ret)
+		return urma_ret;
+
+	if (verify_ret == URMA_EEXIST)
+		return URMA_SUCCESS;
+
+	ret = urma_cmd_set_jetty_opt(jetty, opt, buf, len, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("set jetty opt failed, ret = %d.\n", ret);
+		return URMA_FAIL;
+	}
+
+	return URMA_SUCCESS;
+}
+
+void udma_u_get_jetty_field(struct udma_u_jetty_queue *sq, uint64_t opt, void *buf)
+{
+	switch (opt) {
+	case URMA_JFS_SQE_BASE_ADDR:
+		*((uint64_t *)buf) = (uint64_t)(uintptr_t)sq->qbuf;
+		break;
+	case URMA_JFS_DB_ADDR:
+		*((uint64_t *)buf) = (uint64_t)(uintptr_t)(sq->dwqe_addr + UDMA_DOORBELL_OFFSET);
+		break;
+	case URMA_JFS_DB_STATUS:
+		*((uint8_t *)buf) = sq->db_status;
+		break;
+	case URMA_JFS_PI_TYPE:
+		*((uint8_t *)buf) = sq->pi_type;
+		break;
+	default:
+		break;
+	}
+}
+
+void udma_u_get_jfs_cfg_field(urma_jfs_cfg_t *jfs_cfg, uint64_t opt, void *buf)
+{
+	switch (opt) {
+	case URMA_JFS_DEPTH:
+		*((uint32_t *)buf) = jfs_cfg->depth;
+		break;
+	case URMA_JFS_FLAG:
+		*((uint32_t *)buf) = jfs_cfg->flag.value;
+		break;
+	case URMA_JFS_TRANS_MODE:
+		*((uint32_t *)buf) = jfs_cfg->trans_mode;
+		break;
+	case URMA_JFS_PRIORITY:
+		*((uint8_t *)buf) = jfs_cfg->priority;
+		break;
+	case URMA_JFS_MAX_SGE:
+		*((uint8_t *)buf) = jfs_cfg->max_sge;
+		break;
+	case URMA_JFS_MAX_RSGE:
+		*((uint8_t *)buf) = jfs_cfg->max_rsge;
+		break;
+	case URMA_JFS_MAX_INLINE_DATA:
+		*((uint32_t *)buf) = jfs_cfg->max_inline_data;
+		break;
+	case URMA_JFS_RNR_RETRY:
+		*((uint8_t *)buf) = jfs_cfg->rnr_retry;
+		break;
+	case URMA_JFS_ERR_TIMEOUT:
+		*((uint8_t *)buf) = jfs_cfg->err_timeout;
+		break;
+	case URMA_JFS_BIND_JFC:
+		*((uint64_t *)buf) = (uint64_t)(uintptr_t)jfs_cfg->jfc;
+		break;
+	case URMA_JFS_USER_CTX:
+		*((uint64_t *)buf) = jfs_cfg->user_ctx;
+		break;
+	default:
+		break;
+	}
+}
+
+static void udma_u_get_jetty_cfg_field(urma_jetty_cfg_t *jetty_cfg, uint64_t opt, void *buf)
+{
+	switch (opt) {
+	case URMA_JETTY_ID:
+		*((uint32_t *)buf) = jetty_cfg->id;
+		break;
+	case URMA_JETTY_FLAG:
+		*((uint32_t *)buf) = jetty_cfg->flag.value;
+		break;
+	case URMA_JETTY_BIND_JFR:
+		*((uint64_t *)buf) =
+			(uint64_t)(uintptr_t)jetty_cfg->shared.jfr;
+		break;
+	case URMA_JETTY_BIND_RX_JFC:
+		*((uint64_t *)buf) =
+			(uint64_t)(uintptr_t)jetty_cfg->shared.jfc;
+		break;
+	case URMA_JETTY_BIND_JTG:
+		*((uint64_t *)buf) = (uint64_t)(uintptr_t)jetty_cfg->jetty_grp;
+		break;
+	case URMA_JETTY_USER_CTX:
+		*((uint64_t *)buf) = jetty_cfg->user_ctx;
+		break;
+	default:
+		udma_u_get_jfs_cfg_field(&jetty_cfg->jfs_cfg, opt, buf);
+		break;
+	}
+}
+
+urma_status_t udma_u_get_jetty_opt(urma_jetty_t *jetty, uint64_t opt, void *buf, uint32_t len)
+{
+	struct udma_u_jetty *udma_jetty = to_udma_u_jetty(jetty);
+	urma_cmd_udrv_priv_t udata;
+	urma_status_t urma_ret;
+	int ret;
+
+	urma_ret = udma_u_verify_jetty_opt(opt, buf, len, JETTY_MODE, PERM_R);
+	if (urma_ret && urma_ret != URMA_EEXIST)
+		return urma_ret;
+
+	if (urma_ret != URMA_EEXIST) {
+		ret = urma_cmd_get_jetty_opt(jetty, opt, buf, len, &udata);
+		if (ret) {
+			UDMA_LOG_ERR("set jetty opt failed, ret = %d.\n", ret);
+			return URMA_FAIL;
+		}
+	}
+
+	udma_u_get_jetty_field(&udma_jetty->sq, opt, buf);
+	udma_u_get_jetty_cfg_field(&jetty->jetty_cfg, opt, buf);
+
+	return URMA_SUCCESS;
+}
+
+static int udma_u_active_jetty_after(urma_jetty_t *jetty)
+{
+	struct udma_u_jetty *udma_jetty = to_udma_u_jetty(jetty);
+	int ret;
+
+	udma_jetty->sq.db.type = UDMA_MMAP_JETTY_DSQE;
+
+	ret = udma_u_alloc_db(jetty->urma_ctx, &udma_jetty->sq.db);
+	if (ret)
+		return ret;
+
+	udma_jetty->sq.dwqe_addr = (void *)udma_jetty->sq.db.addr;
+
+	ret = insert_jetty_node(udma_jetty->sq.ctx, udma_jetty);
+	if (ret)
+		udma_u_free_db(jetty->urma_ctx, &udma_jetty->sq.db);
+
+	return ret;
+}
+
+urma_status_t udma_u_active_jetty(urma_jetty_t *jetty)
+{
+	struct udma_u_jetty *udma_jetty = to_udma_u_jetty(jetty);
+	urma_jetty_cfg_t *cfg = &jetty->jetty_cfg;
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	ret = udma_u_active_jetty_prepare(jetty, cfg);
+	if (ret)
+		return ret;
+
+	udma_jetty->jetty_type = UDMA_URMA_EX_JETTY_TYPE;
+	ret = exec_jetty_active_cmd(udma_jetty, cfg);
+	if (ret) {
+		UDMA_LOG_ERR("exec active jetty cmd failed, ret = %d.\n", ret);
+		goto err_active_jetty_cmd;
+	}
+
+	ret = udma_u_active_jetty_after(jetty);
+	if (ret)
+		goto err_active_jetty_after;
+
+	return URMA_SUCCESS;
+err_active_jetty_after:
+	urma_cmd_deactive_jetty(jetty, &udata);
+err_active_jetty_cmd:
+	udma_u_delete_sq(&udma_jetty->sq);
+	remove_jetty_from_grp(udma_jetty);
+
+	return URMA_FAIL;
+}
+
+urma_status_t udma_u_deactive_jetty(urma_jetty_t *jetty)
+{
+	urma_cmd_udrv_priv_t udata;
+	int ret;
+
+	ret = udma_u_delete_jetty_prepare(jetty);
+	if (ret)
+		return URMA_FAIL;
+
+	ret = urma_cmd_deactive_jetty(jetty, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("deactive jetty failed, ret = %d.\n", ret);
+		return URMA_FAIL;
+	}
+
+	udma_u_free_jetty_prepare(jetty);
+
+	return URMA_SUCCESS;
+}
diff --git a/src/urma/hw/udma/udma_u_jetty.h b/src/urma/hw/udma/udma_u_jetty.h
index d40756b..aebfdb0 100644
--- a/src/urma/hw/udma/udma_u_jetty.h
+++ b/src/urma/hw/udma/udma_u_jetty.h
@@ -16,6 +16,28 @@
 #define INVALID_TPN UINT32_MAX
 #define MAX_JETTY_IN_GRP 32
 
+enum udma_u_set_get_jetty_opt_perm {
+	PERM_R = 1,
+	PERM_W = 1 << 1,
+};
+
+enum udma_u_set_get_jetty_opt_ignore_attr {
+	JFS_IGNORE = 1,
+	JETTY_IGNORE = 1 << 1,
+};
+
+enum udma_u_set_get_jetty_opt_mode {
+	JFS_MODE = 1,
+	JETTY_MODE = 1 << 1,
+};
+
+struct udma_u_jetty_opt_info {
+	uint32_t buf_len;
+	enum udma_u_set_get_jetty_opt_perm perm;
+	enum udma_u_set_get_jetty_opt_ignore_attr ignore_attr;
+	enum udma_u_set_get_jetty_opt_mode mode;
+};
+
 urma_jetty_t *udma_u_create_jetty(urma_context_t *ctx, urma_jetty_cfg_t *cfg);
 urma_status_t udma_u_delete_jetty(urma_jetty_t *jetty);
 urma_status_t udma_u_delete_jetty_batch(urma_jetty_t **jetty, int jetty_cnt, urma_jetty_t **bad_jetty);
@@ -53,4 +75,18 @@ urma_status_t insert_jetty_node(struct udma_u_context *udma_ctx,
 				struct udma_u_jetty *pointer);
 void udma_u_jetty_table_remove(struct udma_u_context *udma_ctx,
 				struct udma_u_jetty *jetty);
+urma_status_t udma_u_verify_jetty_opt(uint64_t opt, void *buf, uint32_t len,
+	enum udma_u_set_get_jetty_opt_mode jetty_mode,
+	enum udma_u_set_get_jetty_opt_perm jetty_perm);
+urma_status_t udma_u_set_jetty_field(struct udma_u_jetty_queue *sq, uint64_t opt, void *buf);
+void udma_u_get_jetty_field(struct udma_u_jetty_queue *sq, uint64_t opt, void *buf);
+urma_status_t udma_u_alloc_jetty(urma_context_t *ctx, urma_jetty_cfg_t *cfg,
+				urma_jetty_t **jetty);
+urma_status_t udma_u_set_jetty_opt(urma_jetty_t *jetty, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_get_jetty_opt(urma_jetty_t *jetty, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_active_jetty(urma_jetty_t *jetty);
+urma_status_t udma_u_deactive_jetty(urma_jetty_t *jetty);
+urma_status_t udma_u_free_jetty(urma_jetty_t *jetty);
+void udma_u_get_jfs_cfg_field(urma_jfs_cfg_t *jfs_cfg, uint64_t opt, void *buf);
+
 #endif /* __UDMA_U_JETTY_H__ */
diff --git a/src/urma/hw/udma/udma_u_jfc.c b/src/urma/hw/udma/udma_u_jfc.c
index 7455542..166ff43 100644
--- a/src/urma/hw/udma/udma_u_jfc.c
+++ b/src/urma/hw/udma/udma_u_jfc.c
@@ -53,6 +53,21 @@ static int udma_u_jfc_cmd(urma_context_t *ctx, struct udma_u_jfc *jfc,
 	return urma_cmd_create_jfc(ctx, &jfc->base, cfg, &udata);
 }
 
+static int udma_u_active_jfc_cmd(struct udma_u_jfc *jfc)
+{
+	struct udma_create_jfc_ucmd cmd = {};
+	urma_cmd_udrv_priv_t udata = {};
+
+	cmd.buf_addr = (uintptr_t)jfc->cq.qbuf;
+	cmd.db_addr = (uintptr_t)jfc->sw_db;
+	cmd.buf_len = jfc->cq.qbuf_size;
+	cmd.is_hugepage = jfc->cq.hugepage != NULL;
+
+	udma_u_set_udata(&udata, &cmd, sizeof(cmd), NULL, 0);
+
+	return urma_cmd_active_jfc(&jfc->base, &udata);
+}
+
 static int udma_u_create_cq(struct udma_u_jetty_queue *cq, urma_jfc_cfg_t *cfg)
 {
 	uint32_t depth;
@@ -137,24 +152,278 @@ err_create_cq:
 	return NULL;
 }
 
-urma_status_t udma_u_delete_jfc(urma_jfc_t *jfc)
+urma_status_t udma_u_alloc_jfc(urma_context_t *ctx, urma_jfc_cfg_t *cfg, urma_jfc_t **jfc)
 {
-	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfc->urma_ctx);
-	struct udma_u_jfc *udma_jfc = to_udma_u_jfc(jfc);
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(ctx);
+	urma_cmd_udrv_priv_t udata = {};
+	struct udma_u_jfc *ujfc;
+	int ret;
 
-	if (urma_cmd_delete_jfc(jfc)) {
-		UDMA_LOG_ERR("ubcore delete jfc failed.\n");
+	if (udma_u_check_jfc_cfg(ctx, cfg))
+		return URMA_EINVAL;
+		
+	ujfc = (struct udma_u_jfc *)calloc(1, sizeof(*ujfc));
+	if (!ujfc) {
+		UDMA_LOG_ERR("failed to alloc user udma jfc memory.\n");
+		return URMA_ENOMEM;
+	}
+		
+	ret = urma_cmd_alloc_jfc(ctx, cfg, &ujfc->base, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("udma jfc failed to create urma cmd.\n");
+		free(ujfc);
+		return ret;
+	}
+
+	ujfc->cq.ctx = udma_ctx;
+	*jfc = &ujfc->base;
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_active_jfc(urma_jfc_t *jfc)
+{
+	struct udma_u_jfc *ujfc = to_udma_u_jfc(jfc);
+	int ret;
+
+	if (udma_u_create_cq(&ujfc->cq, &jfc->jfc_cfg)) {
+		UDMA_LOG_ERR("failed to create cq.\n");
 		return URMA_FAIL;
 	}
 
-	if (!udma_jfc->mode) {
-		udma_u_free_sw_db(udma_ctx, udma_jfc->sw_db, UDMA_JFC_TYPE_DB);
+	ujfc->sw_db = (uint32_t *)udma_u_alloc_sw_db(ujfc->cq.ctx, UDMA_JFC_TYPE_DB);
+	if (!ujfc->sw_db) {
+		UDMA_LOG_ERR("failed to alloc user jfc sw db.\n");
+		goto err_alloc_sw_db;
+	}
+
+	ret = udma_u_active_jfc_cmd(ujfc);
+	if (ret != 0) {
+		UDMA_LOG_ERR("udma jfc failed to create urma cmd.\n");
+		goto err_create_jfc;
+	}
+
+	ujfc->cq_shift = align_power2(ujfc->cq.baseblk_cnt);
+	ujfc->cq.idx = ujfc->base.jfc_id.id;
+	ujfc->arm_sn = 1;
+
+	return URMA_SUCCESS;
+
+err_create_jfc:
+	udma_u_free_sw_db(ujfc->cq.ctx, ujfc->sw_db, UDMA_JFC_TYPE_DB);
+err_alloc_sw_db:
+	udma_u_delete_cq(&ujfc->cq);
+
+	return URMA_FAIL;
+}
+
+static int udma_u_set_jfc_depth(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{
+	urma_cmd_udrv_priv_t udata = {};
+
+	jfc->jfc_cfg.depth = *(uint32_t *)buf;
+
+	return urma_cmd_set_jfc_opt(jfc, opt, buf, len, &udata);
+}
+
+static int udma_u_set_jfc_ceqn(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{
+	struct urma_device *dev = jfc->urma_ctx->dev;
+	urma_device_cap_t dev_cap = dev->sysfs_dev->dev_attr.dev_cap;
+	uint32_t ceqn_to_set = *(uint32_t *)buf;
+	urma_cmd_udrv_priv_t udata = {};
+
+	if (ceqn_to_set >= dev_cap.ceq_cnt) {
+		UDMA_LOG_ERR("fail to set ceqn,invalid ceqn = %u, cap ceq cnt = %u.\n", ceqn_to_set, dev_cap.ceq_cnt);
+		return EINVAL;
+	}
+
+	return urma_cmd_set_jfc_opt(jfc, opt, buf, len, &udata);
+}
+
+static int udma_u_set_jfc_flag(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{
+	urma_jfc_flag_t *flag = (urma_jfc_flag_t *)buf;
+	urma_cmd_udrv_priv_t udata = {};
+
+	jfc->jfc_cfg.flag = *flag;
+
+	return urma_cmd_set_jfc_opt(jfc, opt, buf, len, &udata);
+}
+
+static int udma_u_set_jfc_cqe_base_addr(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{
+	struct udma_u_jfc *ujfc = to_udma_u_jfc(jfc);
+	urma_cmd_udrv_priv_t udata = {};
+
+	ujfc->cq.qbuf = (void*)*(uint64_t *)buf;
+	ujfc->cq.cstm = true;
+
+	return urma_cmd_set_jfc_opt(jfc, opt, buf, len, &udata);
+}
+
+static int udma_u_set_jfc_id(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{   
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	ret = urma_cmd_set_jfc_opt(jfc, opt, buf, len, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("failed to set jfc id, id in invalid range or used, ret = %d.\n", ret);
+		return ret;
+	}
+
+	return ret;
+}
+
+static int udma_u_get_jfc_opt_from_kernel(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{
+	urma_cmd_udrv_priv_t udata = {};
+
+	return urma_cmd_get_jfc_opt(jfc, opt, buf, len, &udata);
+}
+
+static int udma_u_get_jfc_flag(urma_jfc_t *jfc, void *buf)
+{
+	urma_jfc_flag_t flag = jfc->jfc_cfg.flag;
+
+	(void)memcpy(buf, &flag, sizeof(flag));
+
+	return 0;
+}
+
+static int udma_u_get_jfc_cqe_base_addr(urma_jfc_t *jfc, void *buf)
+{
+	struct udma_u_jfc *ujfc = to_udma_u_jfc(jfc);
+
+	*(uint64_t*)buf = (uint64_t)(uintptr_t)ujfc->cq.qbuf;
+
+	return URMA_SUCCESS;
+}
+static int udma_u_get_jfc_db_addr(urma_jfc_t *jfc, void *buf)
+{
+	struct udma_u_jfc *ujfc = to_udma_u_jfc(jfc);
+
+	*((uint64_t *)buf) = (uint64_t)(uintptr_t)ujfc->sw_db;
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_set_jfc_opt(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{
+	switch (opt) {
+	case URMA_JFC_DEPTH:
+		return udma_u_set_jfc_depth(jfc, opt, buf, len);
+	case URMA_JFC_CEQN:
+		return udma_u_set_jfc_ceqn(jfc, opt, buf, len);
+	case URMA_JFC_FLAG:
+		return udma_u_set_jfc_flag(jfc, opt, buf, len);
+	case URMA_JFC_BIND_JFCE:
+		break;
+	 case URMA_JFC_USER_CTX:
+		break;
+	case URMA_JFC_CQE_BASE_ADDR:
+		return udma_u_set_jfc_cqe_base_addr(jfc, opt, buf, len);
+	case URMA_JFC_ID:
+		return udma_u_set_jfc_id(jfc, opt, buf, len);
+	case URMA_JFC_DB_STATUS:
+		break;
+	case URMA_JFC_PI_TYPE:
+		break;
+	default:
+		return URMA_FAIL;
+	}
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_get_jfc_opt(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len)
+{
+	switch (opt) {
+	case URMA_JFC_DEPTH:
+		return udma_u_get_jfc_opt_from_kernel(jfc, opt, buf, len);
+	case URMA_JFC_CEQN:
+		return udma_u_get_jfc_opt_from_kernel(jfc, opt, buf, len);
+	case URMA_JFC_FLAG:
+		return udma_u_get_jfc_flag(jfc, buf);
+	case URMA_JFC_BIND_JFCE:
+		*(uint64_t *)buf = (uintptr_t)jfc->jfc_cfg.jfce;
+		break;
+	case URMA_JFC_USER_CTX:
+		*(uint64_t *)buf = jfc->jfc_cfg.user_ctx;
+		break;
+	case URMA_JFC_CQE_BASE_ADDR:
+		return udma_u_get_jfc_cqe_base_addr(jfc, buf);
+	case URMA_JFC_ID:
+		*(uint32_t *)buf = jfc->jfc_id.id;
+		break;
+	case URMA_JFC_DB_ADDR:
+		return udma_u_get_jfc_db_addr(jfc, buf);
+	case URMA_JFC_DB_STATUS:
+		break;
+	case URMA_JFC_PI:
+		return udma_u_get_jfc_opt_from_kernel(jfc, opt, buf, len);
+	case URMA_JFC_PI_TYPE:
+		break;
+	case URMA_JFC_CI:
+		return udma_u_get_jfc_opt_from_kernel(jfc, opt, buf, len);
+	default:
+		return URMA_FAIL;
+	}
+	return URMA_SUCCESS;
+}
+
+static void udma_u_free_jfc_prepare(struct udma_u_jfc *udma_jfc, struct udma_u_context *udma_ctx)
+{
+	if (udma_jfc->cq.cstm == false) {
 		udma_u_free_queue_buf(&udma_jfc->cq);
 	}
+	udma_u_free_sw_db(udma_ctx, udma_jfc->sw_db, UDMA_JFC_TYPE_DB);
 
 	if (!udma_jfc->cq.lock_free)
 		(void)pthread_spin_destroy(&udma_jfc->cq.lock);
+}
+
+urma_status_t udma_u_deactive_jfc(urma_jfc_t *jfc)
+{
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfc->urma_ctx);
+	struct udma_u_jfc *udma_jfc = to_udma_u_jfc(jfc);
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	ret = urma_cmd_deactive_jfc(jfc, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("ubcore deactive jfc failed, errcode:%d.\n", ret);
+		return URMA_FAIL;
+	}
+	udma_u_free_jfc_prepare(udma_jfc, udma_ctx);
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_free_jfc(urma_jfc_t *jfc)
+{
+	struct udma_u_jfc *udma_jfc = to_udma_u_jfc(jfc);
+	urma_cmd_udrv_priv_t udata = {};
+	
+	if (urma_cmd_free_jfc(jfc, &udata)) {
+		UDMA_LOG_ERR("ubcore free jfc failed.\n");
+		return URMA_FAIL;
+	}
+	free(udma_jfc);
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_delete_jfc(urma_jfc_t *jfc)
+{
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfc->urma_ctx);
+	struct udma_u_jfc *udma_jfc = to_udma_u_jfc(jfc);
+
+	if (urma_cmd_delete_jfc(jfc)) {
+		UDMA_LOG_ERR("ubcore delete jfc failed.\n");
+		return URMA_FAIL;
+	}
 
+	udma_u_free_jfc_prepare(udma_jfc, udma_ctx);
 	free(udma_jfc);
 
 	return URMA_SUCCESS;
diff --git a/src/urma/hw/udma/udma_u_jfc.h b/src/urma/hw/udma/udma_u_jfc.h
index d10e4ae..1b99a4a 100644
--- a/src/urma/hw/udma/udma_u_jfc.h
+++ b/src/urma/hw/udma/udma_u_jfc.h
@@ -59,6 +59,12 @@ static inline void *get_u_buf_entry(struct udma_u_jetty_queue *cq, uint32_t n)
 }
 
 urma_jfc_t *udma_u_create_jfc(urma_context_t *ctx, urma_jfc_cfg_t *cfg);
+urma_status_t udma_u_alloc_jfc(urma_context_t *ctx, urma_jfc_cfg_t *cfg, urma_jfc_t **jfc);
+urma_status_t udma_u_set_jfc_opt(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_active_jfc(urma_jfc_t *jfc);
+urma_status_t udma_u_get_jfc_opt(urma_jfc_t *jfc, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_deactive_jfc(urma_jfc_t *jfc);
+urma_status_t udma_u_free_jfc(urma_jfc_t *jfc);
 urma_status_t udma_u_delete_jfc(urma_jfc_t *jfc);
 int udma_u_poll_jfc(urma_jfc_t *jfc, int cr_cnt, urma_cr_t *cr);
 int udma_u_wait_jfc(urma_jfce_t *jfce, uint32_t jfc_cnt, int time_out,
diff --git a/src/urma/hw/udma/udma_u_jfr.c b/src/urma/hw/udma/udma_u_jfr.c
index fdd06ca..f479878 100644
--- a/src/urma/hw/udma/udma_u_jfr.c
+++ b/src/urma/hw/udma/udma_u_jfr.c
@@ -18,9 +18,10 @@ int udma_u_verify_jfr_param(urma_context_t *ctx, urma_jfr_cfg_t *cfg)
 {
 	urma_device_cap_t *cap = &ctx->dev->sysfs_dev->dev_attr.dev_cap;
 
-	if (!cfg->max_sge || !cfg->depth || cfg->depth > cap->max_jfr_depth ||
-	    cfg->max_sge > cap->max_jfr_sge) {
-		UDMA_LOG_ERR("Invalid jfr param, depth = %u, max_sge = %u.\n",
+	if (!cfg->max_sge || !cfg->depth ||
+		roundup_pow_of_two(cfg->depth) > cap->max_jfr_depth ||
+	    roundup_pow_of_two(cfg->max_sge) > cap->max_jfr_sge) {
+		UDMA_LOG_ERR("invalid jfr param, depth = %u, max_sge = %u.\n",
 			     cfg->depth, cfg->max_sge);
 		return EINVAL;
 	}
@@ -236,16 +237,21 @@ static void udma_u_jfr_table_remove(struct udma_u_context *udma_ctx,
 		(void)pthread_rwlock_unlock(&udma_ctx->jfr_table_lock);
 		return;
 	}
+
+	if (!udma_ctx->jfr_table[table_id].jfr_array[jfr->rq.idx & mask]) {
+		(void)pthread_rwlock_unlock(&udma_ctx->jfr_table_lock);
+		return;
+	}
+	udma_ctx->jfr_table[table_id].jfr_array[jfr->rq.idx & mask] = NULL;
+
 	if (!--udma_ctx->jfr_table[table_id].refcnt) {
 		free(udma_ctx->jfr_table[table_id].jfr_array);
 		udma_ctx->jfr_table[table_id].jfr_array = NULL;
-	} else {
-		udma_ctx->jfr_table[table_id].jfr_array[jfr->rq.idx & mask] = NULL;
 	}
 	(void)pthread_rwlock_unlock(&udma_ctx->jfr_table_lock);
 }
 
-static void udma_u_free_jfr(urma_jfr_t *jfr)
+static void udma_u_free_jfr_prepare(urma_jfr_t *jfr)
 {
 	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfr->urma_ctx);
 	struct udma_u_jfr *udma_jfr = to_udma_u_jfr(jfr);
@@ -266,7 +272,12 @@ static void udma_u_free_jfr(urma_jfr_t *jfr)
 		(void)pthread_spin_destroy(&udma_jfr->lock);
 
 	udma_u_jfr_table_remove(udma_ctx, udma_jfr);
-	free(udma_jfr);
+}
+
+static void udma_u_free_jfr_detail(urma_jfr_t *jfr)
+{
+	udma_u_free_jfr_prepare(jfr);
+	free(jfr);
 }
 
 urma_status_t udma_u_delete_jfr(urma_jfr_t *jfr)
@@ -280,8 +291,7 @@ urma_status_t udma_u_delete_jfr(urma_jfr_t *jfr)
 		UDMA_LOG_ERR("urma cmd delete jfr failed, ret = %d.\n", ret);
 		goto delete_err;
 	}
-
-	udma_u_free_jfr(jfr);
+	udma_u_free_jfr_detail(jfr);
 
 	return URMA_SUCCESS;
 
@@ -315,7 +325,7 @@ urma_status_t udma_u_delete_jfr_batch(urma_jfr_t **jfr, int jfr_cnt, urma_jfr_t
 	}
 
 	for (i = 0; i < jfr_cnt; i++)
-		udma_u_free_jfr(jfr[i]);
+		udma_u_free_jfr_detail(jfr[i]);
 
 	return URMA_SUCCESS;
 
@@ -541,3 +551,348 @@ urma_target_jetty_t *udma_u_import_jfr_ex(urma_context_t *ctx,
 
 	return &tjfr->urma_tjetty;
 }
+
+static int udma_u_active_jfr_cmd(struct udma_u_jfr *jfr)
+{
+	struct udma_create_jetty_ucmd cmd = {};
+	struct udma_create_jfr_resp resp = {};
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	cmd.buf_addr = (uintptr_t)jfr->rq.qbuf;
+	cmd.buf_len = jfr->rq.qbuf_size;
+
+	cmd.db_addr = (uintptr_t)jfr->sw_db;
+	cmd.jfr_sleep_buf = (uintptr_t)jfr->long_sleeptime;
+	cmd.idx_addr = (uintptr_t)jfr->idx_que.buf.buf;
+	cmd.idx_len = jfr->idx_que.buf.length;
+	cmd.jetty_addr = (uintptr_t)&jfr->rq;
+	cmd.non_pin = jfr->rq.cstm;
+	cmd.is_hugepage = jfr->rq.hugepage != NULL;
+
+	udma_u_set_udata(&udata, &cmd, sizeof(cmd), &resp, sizeof(resp));
+	ret = urma_cmd_active_jfr(&jfr->base, &udata);
+	if (ret)
+		return ret;
+
+	jfr->cap_flags = resp.jfr_caps;
+	jfr->rq.idx = jfr->base.jfr_id.id;
+
+	return 0;
+}
+
+urma_status_t udma_u_alloc_jfr(urma_context_t *ctx, urma_jfr_cfg_t *cfg, urma_jfr_t **jfr)
+{
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(ctx);
+	urma_cmd_udrv_priv_t udata = {};
+	struct udma_u_jfr *udma_jfr;
+	int ret;
+	
+	udma_jfr = (struct udma_u_jfr *)calloc(1, sizeof(*udma_jfr));
+	if (!udma_jfr) {
+		UDMA_LOG_ERR("calloc jfr is NULL.\n");
+		return URMA_ENOMEM;
+	}
+
+	ret = urma_cmd_alloc_jfr(ctx, cfg, &udma_jfr->base, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("urma cmd alloc jfr failed, ret = %d.\n", ret);
+		free(udma_jfr);
+		udma_jfr = NULL;
+		return URMA_FAIL;
+	}
+	*jfr = &udma_jfr->base;
+	udma_jfr->rq.ctx = udma_ctx;
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_free_jfr(urma_jfr_t *jfr)
+{
+	urma_cmd_udrv_priv_t udata = {};
+
+	if (urma_cmd_free_jfr(jfr, &udata)) {
+		UDMA_LOG_ERR("failed to free user jfr.\n");
+		return URMA_FAIL;
+	}
+
+	free(jfr);
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_active_jfr(urma_jfr_t *jfr)
+{
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfr->urma_ctx);
+	struct udma_u_jfr *udma_jfr = to_udma_u_jfr(jfr);
+	urma_jfr_cfg_t *jfr_cfg = &jfr->jfr_cfg;
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	if (udma_u_verify_jfr_param(jfr->urma_ctx, jfr_cfg))
+		return URMA_EINVAL;
+
+	udma_u_init_jfr_param(udma_jfr, jfr_cfg);
+	if (!udma_jfr->lock_free &&
+	    pthread_spin_init(&udma_jfr->lock, PTHREAD_PROCESS_PRIVATE))
+		return URMA_FAIL;
+
+	if (udma_u_alloc_jfr_idx_que(udma_jfr)) {
+		UDMA_LOG_ERR("failed to alloc jfr idx que.\n");
+		goto err_alloc_idx;
+	}
+
+	if (udma_u_create_rq(udma_ctx, udma_jfr)) {
+		UDMA_LOG_ERR("failed to create jfr rqe buf.\n");
+		goto err_create_rq;
+	}
+
+	if (!udma_jfr->swdb_cstm) {
+		udma_jfr->sw_db = (uint32_t *)udma_u_alloc_sw_db(udma_ctx, UDMA_JFR_TYPE_DB);
+		if (!udma_jfr->sw_db) {
+			UDMA_LOG_ERR("failed to alloc sw db.\n");
+			goto err_alloc_sw_db;
+		}
+	
+		udma_jfr->long_sleeptime = (bool *)udma_u_alloc_sw_db(udma_ctx, UDMA_JFR_PAYLOAD);
+		if (!udma_jfr->long_sleeptime) {
+			UDMA_LOG_ERR("failed to alloc sw db for payload\n");
+			goto err_alloc_jfr_sleep_buf;
+		}
+		*udma_jfr->long_sleeptime = false;
+	}
+
+	ret = udma_u_active_jfr_cmd(udma_jfr);
+	if (ret) {
+		UDMA_LOG_ERR("failed to active jfr cmd, ret = %d.\n", ret);
+		goto err_active_cmd;
+	}
+
+	if (udma_u_insert_jfr_node(udma_ctx, udma_jfr))
+		goto err_insert_node;
+
+	return URMA_SUCCESS;
+
+err_insert_node:
+	urma_cmd_deactive_jfr(&udma_jfr->base, &udata);
+err_active_cmd:
+	if (!udma_jfr->swdb_cstm)
+		udma_u_free_sw_db(udma_ctx, (uint32_t *)udma_jfr->long_sleeptime, UDMA_JFR_PAYLOAD);
+err_alloc_jfr_sleep_buf:
+	if (!udma_jfr->swdb_cstm)
+		udma_u_free_sw_db(udma_ctx, udma_jfr->sw_db, UDMA_JFR_TYPE_DB);
+err_alloc_sw_db:
+	udma_u_free_queue_buf(&udma_jfr->rq);
+err_create_rq:
+	udma_u_free_idx_que(&udma_jfr->idx_que);
+err_alloc_idx:
+	if (!udma_jfr->lock_free)
+		pthread_spin_destroy(&udma_jfr->lock);
+
+	return URMA_FAIL;
+}
+
+static int udma_u_set_jfr_param(urma_jfr_t *jfr, uint64_t opt, void *buf)
+{
+	struct udma_u_jfr *udma_jfr = to_udma_u_jfr(jfr);
+
+	switch (opt) {
+	case URMA_JFR_DEPTH:
+	case URMA_JFR_FLAG:
+	case URMA_JFR_TRANS_MODE:
+	case URMA_JFR_MAX_SGE:
+	case URMA_JFR_MIN_RNR_TIMER:
+	case URMA_JFR_BIND_JFC:
+	case URMA_JFR_TOKEN_VALUE:
+		break;
+	case URMA_JFR_USER_CTX:
+		return URMA_EEXIST;
+	case URMA_JFR_RQE_BASE_ADDR:
+		udma_jfr->rq.qbuf = (void *)(uintptr_t)(*((uint64_t *)buf));
+		if (!(udma_jfr->rq.qbuf)) {
+			UDMA_LOG_ERR("failed to set rq addr.\n");
+			return URMA_EINVAL;
+		}
+		udma_jfr->rq.cstm = true;
+		break;
+	case URMA_JFR_ID:
+		udma_jfr->rq.idx = *((uint32_t *)buf);
+		break;
+	case URMA_JFR_DB_ADDR:
+		udma_jfr->sw_db = (void *)(uintptr_t)(*((uint64_t *)buf));
+		if (!(udma_jfr->sw_db)) {
+			UDMA_LOG_ERR("failed to set db addr.\n");
+			return URMA_EINVAL;
+		}
+		udma_jfr->swdb_cstm = true;
+		break;
+	case URMA_JFR_DB_STATUS:
+	case URMA_JFR_PI_TYPE:
+		/* TODO */
+		break;
+	default:
+		UDMA_LOG_ERR("invalid param, opt=%lu.\n", opt);
+		return URMA_EINVAL;
+	}
+
+	return 0;
+}
+
+static int udma_u_get_jfr_param(urma_jfr_t *jfr, uint64_t opt, void *buf)
+{
+	struct udma_u_jfr *udma_jfr = to_udma_u_jfr(jfr);
+
+	switch (opt) {
+	case URMA_JFR_DEPTH:
+		*((uint32_t *)buf) = (jfr->jfr_cfg.depth);
+		break;
+	case URMA_JFR_FLAG:
+		*((uint32_t *)buf) = (jfr->jfr_cfg.flag.value);
+		break;
+	case URMA_JFR_TRANS_MODE:
+		*((uint32_t *)buf) = (jfr->jfr_cfg.trans_mode);
+		break;
+	case URMA_JFR_MAX_SGE:
+		*((uint8_t *)buf) = (jfr->jfr_cfg.max_sge);
+		break;
+	case URMA_JFR_MIN_RNR_TIMER:
+		*((uint8_t *)buf) = (jfr->jfr_cfg.min_rnr_timer);
+		break;
+	case URMA_JFR_BIND_JFC:
+		*((uint64_t *)buf) = (uint64_t)(uintptr_t)(jfr->jfr_cfg.jfc);
+		break;
+	case URMA_JFR_TOKEN_VALUE:
+		*((uint32_t *)buf) = jfr->jfr_cfg.token_value.token;
+		break;
+	case URMA_JFR_USER_CTX:
+		*((uint64_t *)buf) = jfr->jfr_cfg.user_ctx;
+		break;
+	case URMA_JFR_RQE_BASE_ADDR:
+		*((uint64_t *)buf) = (uint64_t)(uintptr_t)udma_jfr->rq.qbuf;
+		break;
+	case URMA_JFR_ID:
+		*((uint32_t *)buf) = udma_jfr->rq.idx;
+		break;
+	case URMA_JFR_DB_ADDR:
+		*((uint32_t *)buf) = (uint32_t)(uintptr_t)udma_jfr->sw_db;
+		break;
+	case URMA_JFR_DB_STATUS:
+	case URMA_JFR_PI_TYPE:
+		/*TODO*/
+		break;
+	case URMA_JFR_PI:
+	case URMA_JFR_CI:
+		break;
+	default:
+		UDMA_LOG_ERR("invalid param, opt=%lu.\n", opt);
+		return URMA_EINVAL;
+	}
+
+	return 0;
+}
+
+static struct udma_u_jfr_opt_info opt_u_jfr_table[] = {
+	{URMA_JFR_DEPTH, sizeof(uint32_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_FLAG, sizeof(uint32_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_TRANS_MODE, sizeof(uint32_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_MAX_SGE, sizeof(uint8_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_MIN_RNR_TIMER , sizeof(uint8_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_BIND_JFC, sizeof(uint64_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_TOKEN_VALUE, sizeof(uint32_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_USER_CTX, sizeof(uint64_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_RQE_BASE_ADDR, sizeof(uint64_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_ID, sizeof(uint32_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_DB_ADDR, sizeof(uint64_t), PERM_READ},
+	{URMA_JFR_DB_STATUS, sizeof(uint8_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_PI, sizeof(uint16_t), PERM_READ},
+	{URMA_JFR_PI_TYPE, sizeof(uint16_t), PERM_READ | PERM_WRITE},
+	{URMA_JFR_CI, sizeof(uint16_t), PERM_READ},
+};
+
+static int udma_u_check_set_get_jfr_param(uint64_t opt, void *buf, uint32_t len,
+					  enum udma_set_get_jfr_opt_perm perm)
+{
+#define UDMA_JFR_GET_JFR_OPT_CNT 15
+
+	if (!buf) {
+		UDMA_LOG_ERR("invalid param, buf is NULL.\n");
+		return EINVAL;
+	}
+
+	for (size_t i = 0; i < UDMA_JFR_GET_JFR_OPT_CNT; i++) {
+		if ((opt_u_jfr_table[i].opt == opt) && (opt_u_jfr_table[i].buf_len == len) &&
+		    (opt_u_jfr_table[i].perm & perm))
+			return 0;
+	}
+	UDMA_LOG_ERR("invalid param, len = %u, opt = %lu.\n", len, opt);
+
+	return EINVAL;
+}
+
+urma_status_t udma_u_set_jfr_opt(urma_jfr_t *jfr, uint64_t opt, void *buf, uint32_t len)
+{
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	ret = udma_u_check_set_get_jfr_param(opt, buf, len, PERM_WRITE);
+	if (ret)
+		return URMA_EINVAL;
+	
+	ret = udma_u_set_jfr_param(jfr, opt, buf);
+	if (ret == URMA_EEXIST)
+		return URMA_SUCCESS;
+	if (ret)
+		return URMA_FAIL;
+
+	ret = urma_cmd_set_jfr_opt(jfr, opt, buf, len, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("failed to set user jfr, opt=%lu, ret = %d.\n", opt, ret);
+		return URMA_FAIL;
+	}
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_get_jfr_opt(urma_jfr_t *jfr, uint64_t opt, void *buf, uint32_t len)
+{
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	ret = udma_u_check_set_get_jfr_param(opt, buf, len, PERM_READ);
+	if (ret)
+		return URMA_FAIL;
+
+	if (opt != URMA_JFR_USER_CTX) {
+		ret = urma_cmd_get_jfr_opt(jfr, opt, buf, len, &udata);
+		if (ret) {
+			UDMA_LOG_ERR("failed to get jfr opt, opt=%lu, ret = %d.\n", opt, ret);
+			return URMA_FAIL;
+		}
+	}
+	ret = udma_u_get_jfr_param(jfr, opt, buf);
+	if (ret) {
+		UDMA_LOG_ERR("failed to get user jfr param, opt=%lu, ret = %d.\n", opt, ret);
+		return URMA_FAIL;
+	}
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_deactive_jfr(urma_jfr_t *jfr)
+{
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfr->urma_ctx);
+	struct udma_u_jfr *udma_jfr = to_udma_u_jfr(jfr);
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	ret = urma_cmd_deactive_jfr(jfr, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("failed to deactive jfr, ret = %d.\n", ret);
+		udma_u_jfr_table_remove(udma_ctx, udma_jfr);
+		return URMA_FAIL;
+	}
+
+	udma_u_free_jfr_prepare(jfr);
+
+	return URMA_SUCCESS;
+}
\ No newline at end of file
diff --git a/src/urma/hw/udma/udma_u_jfr.h b/src/urma/hw/udma/udma_u_jfr.h
index 9c71a4e..c160ff9 100644
--- a/src/urma/hw/udma/udma_u_jfr.h
+++ b/src/urma/hw/udma/udma_u_jfr.h
@@ -17,6 +17,17 @@
 #define UDMA_U_MIN_JFR_DEPTH 64
 #define UDMA_JFR_LARGE_PACKAGE 4096
 
+enum udma_set_get_jfr_opt_perm {
+	PERM_READ = 1,
+	PERM_WRITE = 1 << 1,
+};
+
+struct udma_u_jfr_opt_info {
+	uint64_t opt;
+	uint32_t buf_len;
+	enum udma_set_get_jfr_opt_perm perm;
+};
+
 static inline bool udma_jfrwq_overflow(struct udma_u_jfr *jfr)
 {
 	return (jfr->rq.pi - jfr->rq.ci) >= jfr->wqe_cnt;
@@ -57,4 +68,10 @@ urma_target_jetty_t *udma_u_import_jfr_ex(urma_context_t *ctx,
 					  urma_token_t *token_value,
 					  urma_active_tp_cfg_t *active_tp_cfg);
 int udma_u_insert_jfr_node(struct udma_u_context *udma_ctx, struct udma_u_jfr *jfr);
+urma_status_t udma_u_alloc_jfr(urma_context_t *ctx, urma_jfr_cfg_t *cfg, urma_jfr_t **jfr);
+urma_status_t udma_u_set_jfr_opt(urma_jfr_t *jfr, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_active_jfr(urma_jfr_t *jfr);
+urma_status_t udma_u_get_jfr_opt(urma_jfr_t *jfr, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_deactive_jfr(urma_jfr_t *jfr);
+urma_status_t udma_u_free_jfr(urma_jfr_t *jfr);
 #endif /* __UDMA_U_JFR_H__ */
diff --git a/src/urma/hw/udma/udma_u_jfs.c b/src/urma/hw/udma/udma_u_jfs.c
index c577bf8..d6e3a76 100644
--- a/src/urma/hw/udma/udma_u_jfs.c
+++ b/src/urma/hw/udma/udma_u_jfs.c
@@ -14,6 +14,7 @@
 #include "udma_u_buf.h"
 #include "udma_u_db.h"
 #include "udma_u_jfc.h"
+#include "udma_u_jetty.h"
 #include "udma_u_jfs.h"
 
 static uint32_t get_ctl_len(uint8_t opcode)
@@ -86,9 +87,66 @@ static bool udma_check_atomic_len(uint32_t len, uint8_t opcode)
 	}
 }
 
+int udma_u_set_sq_by_resp(struct udma_u_jetty_queue *sq,
+			  struct udma_create_jetty_resp *resp)
+{
+	if (!sq->sq_reserved || sq->cstm)
+		return 0;
+
+	if (resp->buf_addr == 0) {
+		UDMA_LOG_ERR("failed to get reserved addr.\n");
+		return EINVAL;
+	}
+
+	sq->qbuf = (void *)(uintptr_t)resp->buf_addr;
+	sq->qbuf_curr = sq->qbuf;
+	sq->qbuf_end = sq->qbuf + sq->qbuf_size;
+
+	return 0;
+}
+
+static int udma_u_exec_jfs_active_cmd(struct udma_u_jfs *jfs, urma_jfs_cfg_t *cfg)
+{
+	struct udma_create_jetty_resp resp = {};
+	struct udma_create_jetty_ucmd cmd = {};
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	if (cfg->priority >= UDMA_MAX_PRIORITY) {
+		UDMA_LOG_ERR("user mode jfs priority is out of range, priority is %u.\n",
+			     cfg->priority);
+		return EINVAL;
+	}
+
+	cmd.buf_addr = (uintptr_t)jfs->sq.qbuf;
+	cmd.buf_len = jfs->sq.qbuf_size;
+	cmd.jetty_addr = (uintptr_t)&jfs->sq;
+	cmd.sqe_bb_cnt = jfs->sq.sqe_bb_cnt;
+	cmd.pi_type = jfs->sq.pi_type;
+	cmd.non_pin = jfs->sq.cstm;
+	cmd.is_hugepage = jfs->sq.hugepage != NULL;
+	cmd.jetty_type = jfs->jfs_type;
+	cmd.jfs_id = jfs->sq.db.id;
+	udma_u_set_udata(&udata, &cmd, (uint32_t)sizeof(cmd), &resp, (uint32_t)sizeof(resp));
+	ret = urma_cmd_active_jfs(&jfs->base, &udata);
+	if (ret != 0) {
+		UDMA_LOG_ERR("failed to urma cmd active jfs, ret is %d.\n", ret);
+		return ret;
+	}
+
+	ret = udma_u_set_sq_by_resp(&jfs->sq, &resp);
+	if (ret != 0) {
+		(void)urma_cmd_deactive_jfs(&jfs->base, NULL);
+		return ret;
+	}
+
+	return ret;
+}
+
 int udma_u_exec_jfs_create_cmd(urma_context_t *ctx, struct udma_u_jfs *jfs,
 			       urma_jfs_cfg_t *cfg)
 {
+	struct udma_create_jetty_resp resp = {};
 	struct udma_create_jetty_ucmd cmd = {};
 	urma_cmd_udrv_priv_t udata = {};
 	int ret;
@@ -103,22 +161,33 @@ int udma_u_exec_jfs_create_cmd(urma_context_t *ctx, struct udma_u_jfs *jfs,
 	cmd.buf_len = jfs->sq.qbuf_size;
 	cmd.jetty_addr = (uintptr_t)&jfs->sq;
 	cmd.sqe_bb_cnt = jfs->sq.sqe_bb_cnt;
-	cmd.pi_type = jfs->pi_type;
+	cmd.pi_type = jfs->sq.pi_type;
 	cmd.non_pin = jfs->sq.cstm;
 	cmd.is_hugepage = jfs->sq.hugepage != NULL;
 	cmd.jetty_type = jfs->jfs_type;
 	cmd.jfs_id = jfs->sq.db.id;
-	udma_u_set_udata(&udata, &cmd, (uint32_t)sizeof(cmd), NULL, 0);
+	udma_u_set_udata(&udata, &cmd, (uint32_t)sizeof(cmd), &resp, (uint32_t)sizeof(resp));
 	ret = urma_cmd_create_jfs(ctx, &jfs->base, cfg, &udata);
-	if (ret != 0)
+	if (ret != 0) {
 		UDMA_LOG_ERR("failed to urma cmd create jfs, ret is %d.\n", ret);
+		return ret;
+	}
+
+	ret = udma_u_set_sq_by_resp(&jfs->sq, &resp);
+	if (ret != 0) {
+		(void)urma_cmd_delete_jfs(&jfs->base);
+		return ret;
+	}
 
 	return ret;
 }
 
 int udma_u_create_sq(struct udma_u_jetty_queue *sq, urma_jfs_cfg_t *cfg)
 {
+	struct udma_u_context *ctx = sq->ctx;
+	uint32_t aligned_size;
 	uint32_t sqe_bb_cnt;
+	bool is_ccu_jetty;
 
 	sq->lock_free = cfg->flag.bs.lock_free;
 
@@ -135,8 +204,14 @@ int udma_u_create_sq(struct udma_u_jetty_queue *sq, urma_jfs_cfg_t *cfg)
 		sqe_bb_cnt = MAX_SQE_BB_NUM;
 	sq->sqe_bb_cnt = sqe_bb_cnt;
 	sq->max_sge_num = cfg->max_sge;
+
+	is_ccu_jetty = ctx->ccu_jetty_max_cnt &&
+		       sq->db.id >= ctx->ccu_jetty_start_id &&
+		       sq->db.id < ctx->ccu_jetty_start_id + ctx->ccu_jetty_max_cnt;
+	aligned_size = is_ccu_jetty ?
+		(UDMA_MIN_CCU_WQEBB_CNT * UDMA_JFS_WQEBB) : UDMA_HW_PAGE_SIZE;
 	if (!udma_u_alloc_queue_buf(sq, sqe_bb_cnt * cfg->depth,
-				    UDMA_JFS_WQEBB, UDMA_HW_PAGE_SIZE, true)) {
+				    UDMA_JFS_WQEBB, aligned_size, true)) {
 		UDMA_LOG_ERR("failed to alloc jfs wqe buf.\n");
 		goto err_alloc_buf;
 	}
@@ -158,6 +233,23 @@ void udma_u_delete_sq(struct udma_u_jetty_queue *sq)
 		(void)pthread_spin_destroy(&sq->lock);
 }
 
+void udma_u_lock_delete_sq(struct udma_u_jetty_queue *sq)
+{
+	if (sq->wrid != NULL) {
+		free(sq->wrid);
+		sq->wrid = NULL;
+	}
+
+	if (sq->qbuf != NULL) {
+		sq->qbuf = NULL;
+		sq->qbuf_curr = NULL;
+		sq->qbuf_end = NULL;
+	}
+
+	if (!sq->lock_free)
+		(void)pthread_spin_destroy(&sq->lock);
+}
+
 urma_jfs_t *udma_u_create_jfs(urma_context_t *ctx, urma_jfs_cfg_t *cfg)
 {
 	struct udma_u_context *udma_ctx = to_udma_u_ctx(ctx);
@@ -175,6 +267,7 @@ urma_jfs_t *udma_u_create_jfs(urma_context_t *ctx, urma_jfs_cfg_t *cfg)
 	}
 
 	jfs->sq.ctx = udma_ctx;
+	jfs->sq.sq_reserved = udma_ctx->sq_reserved;
 	if (udma_u_create_sq(&jfs->sq, cfg)) {
 		UDMA_LOG_ERR("failed to create sq.\n");
 		goto err_create_sq;
@@ -202,8 +295,9 @@ err_create_sq:
 	return NULL;
 }
 
-static void udma_u_free_jfs(urma_jfs_t *jfs)
+static void udma_u_free_jfs_prepare(urma_jfs_t *jfs)
 {
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(jfs->urma_ctx);
 	struct udma_u_jfs *udma_jfs = to_udma_u_jfs(jfs);
 
 	if (!!jfs->jfs_cfg.jfc)
@@ -211,7 +305,12 @@ static void udma_u_free_jfs(urma_jfs_t *jfs)
 
 	udma_u_free_db(jfs->urma_ctx, &udma_jfs->sq.db);
 	udma_u_delete_sq(&udma_jfs->sq);
-	free(udma_jfs);
+}
+
+static void udma_u_free_jfs_detail(urma_jfs_t *jfs)
+{
+	udma_u_free_jfs_prepare(jfs);
+	free(jfs);
 }
 
 urma_status_t udma_u_delete_jfs(urma_jfs_t *jfs)
@@ -219,7 +318,7 @@ urma_status_t udma_u_delete_jfs(urma_jfs_t *jfs)
 	if (urma_cmd_delete_jfs(jfs))
 		return URMA_FAIL;
 
-	udma_u_free_jfs(jfs);
+	udma_u_free_jfs_detail(jfs);
 
 	return URMA_SUCCESS;
 }
@@ -242,7 +341,7 @@ urma_status_t udma_u_delete_jfs_batch(urma_jfs_t **jfs, int jfs_cnt, urma_jfs_t
 		return URMA_FAIL;
 
 	for (i = 0; i < jfs_cnt; i++)
-		udma_u_free_jfs(jfs[i]);
+		udma_u_free_jfs_detail(jfs[i]);
 
 	return URMA_SUCCESS;
 }
@@ -831,10 +930,14 @@ urma_status_t udma_u_post_sq_wr(struct udma_u_context *udma_ctx,
 	if (wr_cnt) {
 		UDMA_TO_DEVICE_BARRIER();
 
-		if (wr_cnt == 1 && dwqe_enable && (sq->pi - sq->ci == 1))
-			udma_write_dsqe(sq, wqe_addr);
-		else
-			udma_update_sq_db(sq);
+		if (UNLIKELY(sq->db_status)) {
+			sq->need_ring_db = true;
+		} else {
+			if (wr_cnt == 1 && dwqe_enable && (sq->pi - sq->ci == 1))
+				udma_write_dsqe(sq, wqe_addr);
+			else
+				udma_update_sq_db(sq);
+		}
 	}
 
 	if (!sq->lock_free)
@@ -916,7 +1019,7 @@ static void copy_from_sq(struct udma_u_jetty_queue *sq, uint32_t wqebb_cnt,
 
 	offset = (sq->ci & (sq->baseblk_cnt - 1)) * UDMA_JFS_WQEBB;
 	field_h = remain > wqebb_cnt ? wqebb_cnt : remain;
-	field_l = wqebb_cnt > field_h ? wqebb_cnt - field_h : 0;
+	field_l = wqebb_cnt - field_h;
 
 	(void)memcpy((void *)tmp_sq, sq->qbuf + offset, field_h * sizeof(struct udma_jfs_wqebb));
 
@@ -1039,3 +1142,188 @@ int udma_u_flush_jfs(urma_jfs_t *jfs, int cr_cnt, urma_cr_t *cr)
 
 	return n_flushed;
 }
+
+urma_status_t udma_u_alloc_jfs(urma_context_t *ctx, urma_jfs_cfg_t *cfg,
+			      urma_jfs_t **jfs)
+{
+	struct udma_u_context *udma_ctx = to_udma_u_ctx(ctx);
+	struct udma_u_jfs *udma_jfs;
+	urma_cmd_udrv_priv_t udata;
+	int ret;
+
+	udma_jfs = (struct udma_u_jfs *)calloc(1, sizeof(struct udma_u_jfs));
+	if (udma_jfs == NULL) {
+		UDMA_LOG_ERR("udma jfs allocation failed.\n");
+		return URMA_ENOMEM;
+	}
+
+	ret = urma_cmd_alloc_jfs(ctx, cfg, &udma_jfs->base, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("alloc jfs failed, ret = %d.\n", ret);
+		free(udma_jfs);
+		return URMA_FAIL;
+	}
+
+	*jfs = &udma_jfs->base;
+	udma_jfs->sq.ctx = udma_ctx;
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_free_jfs(urma_jfs_t *jfs)
+{
+	urma_cmd_udrv_priv_t udata;
+	int ret;
+
+	ret = urma_cmd_free_jfs(jfs, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("free jfs failed, ret = %d.\n", ret);
+		return URMA_FAIL;
+	}
+
+	free(jfs);
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_set_jfs_opt(urma_jfs_t *jfs, uint64_t opt, void *buf, uint32_t len)
+{
+	struct udma_u_jfs *udma_jfs = to_udma_u_jfs(jfs);
+	urma_cmd_udrv_priv_t udata;
+	urma_status_t verify_ret;
+	urma_status_t urma_ret;
+	int ret;
+
+	verify_ret = udma_u_verify_jetty_opt(opt, buf, len, JFS_MODE, PERM_W);
+	if (verify_ret && verify_ret != URMA_EEXIST)
+		return verify_ret;
+
+	if (opt == URMA_JFS_ID)
+		udma_jfs->sq.db.id = *((uint32_t *)(uintptr_t)buf);
+	else {
+		urma_ret = udma_u_set_jetty_field(&udma_jfs->sq, opt, buf);
+		if (urma_ret)
+			return urma_ret;
+	}
+
+	if (verify_ret == URMA_EEXIST)
+		return URMA_SUCCESS;
+
+	ret = urma_cmd_set_jfs_opt(jfs, opt, buf, len, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("set jfs opt failed, ret = %d.\n", ret);
+		return URMA_FAIL;
+	}
+
+	return URMA_SUCCESS;
+}
+
+urma_status_t udma_u_get_jfs_opt(urma_jfs_t *jfs, uint64_t opt, void *buf, uint32_t len)
+{
+	struct udma_u_jfs *udma_jfs = to_udma_u_jfs(jfs);
+	urma_cmd_udrv_priv_t udata;
+	urma_status_t urma_ret;
+	int ret;
+
+	urma_ret = udma_u_verify_jetty_opt(opt, buf, len, JFS_MODE, PERM_R);
+	if (urma_ret && urma_ret != URMA_EEXIST)
+		return urma_ret;
+
+	if (urma_ret != URMA_EEXIST) {
+		ret = urma_cmd_get_jfs_opt(jfs, opt, buf, len, &udata);
+		if (ret) {
+			UDMA_LOG_ERR("set jfs opt failed, ret = %d.\n", ret);
+			return URMA_FAIL;
+		}
+	}
+
+	if (opt == URMA_JFS_ID) {
+		*((uint32_t *)(uintptr_t)buf) = udma_jfs->sq.db.id;
+	} else {
+		udma_u_get_jetty_field(&udma_jfs->sq, opt, buf);
+		udma_u_get_jfs_cfg_field(&jfs->jfs_cfg, opt, buf);
+	}
+
+	return URMA_SUCCESS;
+}
+
+static int udma_u_active_jfs_prepare(urma_jfs_t *jfs, urma_jfs_cfg_t *cfg)
+{
+	struct udma_u_jfs *udma_jfs = to_udma_u_jfs(jfs);
+	int ret;
+
+	if (cfg->trans_mode == URMA_TM_RC) {
+		UDMA_LOG_ERR("jfs not support RC transmode.\n");
+		return EINVAL;
+	}
+
+	udma_jfs->sq.sq_reserved = udma_jfs->sq.sq_reserved;
+	ret = udma_u_create_sq(&udma_jfs->sq, cfg);
+	if (ret)
+		UDMA_LOG_ERR("failed to create sq, ret = %d.\n", ret);
+
+	return ret;
+}
+
+static int udma_u_active_jfs_after(urma_jfs_t *jfs)
+{
+	struct udma_u_jfs *udma_jfs = to_udma_u_jfs(jfs);
+	int ret;
+
+	udma_jfs->sq.db.type = UDMA_MMAP_JETTY_DSQE;
+	udma_jfs->sq.db.id = udma_jfs->base.jfs_id.id;
+	ret = udma_u_alloc_db(jfs->urma_ctx, &udma_jfs->sq.db);
+	if (ret)
+		return ret;
+
+	udma_jfs->sq.dwqe_addr = (void *)udma_jfs->sq.db.addr;
+
+	return ret;
+}
+
+urma_status_t udma_u_active_jfs(urma_jfs_t *jfs)
+{
+	struct udma_u_jfs *udma_jfs = to_udma_u_jfs(jfs);
+	urma_jfs_cfg_t *cfg = &jfs->jfs_cfg;
+	urma_cmd_udrv_priv_t udata = {};
+	int ret;
+
+	ret = udma_u_active_jfs_prepare(jfs, cfg);
+	if (ret)
+		return ret;
+
+	udma_jfs->jfs_type = UDMA_URMA_EX_JETTY_TYPE;
+	ret = udma_u_exec_jfs_active_cmd(udma_jfs, cfg);
+	if (ret) {
+		UDMA_LOG_ERR("exec active jfs cmd failed, ret = %d.\n", ret);
+		goto err_active_jfs_cmd;
+	}
+
+	ret = udma_u_active_jfs_after(jfs);
+	if (ret)
+		goto err_active_jfs_after;
+
+	return URMA_SUCCESS;
+err_active_jfs_after:
+	urma_cmd_deactive_jfs(jfs, &udata);
+err_active_jfs_cmd:
+	udma_u_delete_sq(&udma_jfs->sq);
+
+	return URMA_FAIL;
+}
+
+urma_status_t udma_u_deactive_jfs(urma_jfs_t *jfs)
+{
+	urma_cmd_udrv_priv_t udata;
+	int ret;
+
+	ret = urma_cmd_deactive_jfs(jfs, &udata);
+	if (ret) {
+		UDMA_LOG_ERR("deactive jfs failed, ret = %d.\n", ret);
+		return URMA_FAIL;
+	}
+
+	udma_u_free_jfs_prepare(jfs);
+
+	return URMA_SUCCESS;
+}
diff --git a/src/urma/hw/udma/udma_u_jfs.h b/src/urma/hw/udma/udma_u_jfs.h
index a069545..26e51d7 100644
--- a/src/urma/hw/udma/udma_u_jfs.h
+++ b/src/urma/hw/udma/udma_u_jfs.h
@@ -155,12 +155,15 @@ int udma_u_create_sq(struct udma_u_jetty_queue *sq, urma_jfs_cfg_t *cfg);
 urma_jfs_t *udma_u_create_jfs(urma_context_t *ctx, urma_jfs_cfg_t *cfg);
 urma_status_t udma_u_delete_jfs(urma_jfs_t *jfs);
 void udma_u_delete_sq(struct udma_u_jetty_queue *sq);
+void udma_u_lock_delete_sq(struct udma_u_jetty_queue *sq);
 urma_status_t udma_u_modify_jfs(urma_jfs_t *jfs, urma_jfs_attr_t *jfs_attr);
 urma_status_t udma_u_query_jfs(urma_jfs_t *jfs, urma_jfs_cfg_t *cfg,
 			       urma_jfs_attr_t *attr);
 int udma_u_flush_jfs(urma_jfs_t *jfs, int cr_cnt, urma_cr_t *cr);
 void udma_u_flush_sq(uint32_t local_id, struct udma_u_jetty_queue *sq,
 		     urma_cr_t *cr, bool is_jetty);
+int udma_u_set_sq_by_resp(struct udma_u_jetty_queue *sq,
+			  struct udma_create_jetty_resp *resp);
 int udma_u_exec_jfs_create_cmd(urma_context_t *ctx,
 			       struct udma_u_jfs *jfs,
 			       urma_jfs_cfg_t *cfg);
@@ -171,5 +174,12 @@ urma_status_t udma_u_post_one_wr(struct udma_u_context *udma_ctx,
 				 bool *dwqe_enable);
 void udma_reset_sw_u_jetty_queue(struct udma_u_jetty_queue *sq);
 urma_status_t udma_u_delete_jfs_batch(urma_jfs_t **jfs, int jfs_cnt, urma_jfs_t **bad_jfs);
+urma_status_t udma_u_alloc_jfs(urma_context_t *ctx, urma_jfs_cfg_t *cfg,
+			      urma_jfs_t **jfs);
+urma_status_t udma_u_set_jfs_opt(urma_jfs_t *jfs, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_active_jfs(urma_jfs_t *jfs);
+urma_status_t udma_u_get_jfs_opt(urma_jfs_t *jfs, uint64_t opt, void *buf, uint32_t len);
+urma_status_t udma_u_deactive_jfs(urma_jfs_t *jfs);
+urma_status_t udma_u_free_jfs(urma_jfs_t *jfs);
 
 #endif /* __UDMA_U_JFS_H__ */
diff --git a/src/urma/hw/udma/udma_u_ops.c b/src/urma/hw/udma/udma_u_ops.c
index 4593abe..9dbcfab 100644
--- a/src/urma/hw/udma/udma_u_ops.c
+++ b/src/urma/hw/udma/udma_u_ops.c
@@ -10,6 +10,8 @@
 #include <unistd.h>
 #include <stdio.h>
 #include <sys/mman.h>
+#include <errno.h>
+#include <string.h>
 #include "udma_u_jfc.h"
 #include "udma_u_jfs.h"
 #include "udma_u_jfr.h"
@@ -22,24 +24,46 @@
 #include "udma_u_buf.h"
 #include "udma_u_ops.h"
 
-/* udma_u_xx_ex interface thanks to rdma-core-master/providers/hns/hns_roce_u.c code. */
+static pthread_mutex_t g_sq_reserved_lock = PTHREAD_MUTEX_INITIALIZER;
+static uint64_t g_sq_reserved_refcount = 0;
+static void *g_sq_reserved_va = NULL;
+static uint64_t g_sq_reserved_len = 0;
+
 static urma_ops_t g_udma_ops = {
 	.name = "UDMA_OPS",
 	.create_jfc = udma_u_create_jfc,
 	.modify_jfc = udma_u_modify_jfc,
 	.delete_jfc = udma_u_delete_jfc,
+	.alloc_jfc = udma_u_alloc_jfc,
+	.set_jfc_opt = udma_u_set_jfc_opt,
+	.active_jfc = udma_u_active_jfc,
+	.get_jfc_opt = udma_u_get_jfc_opt,
+	.deactive_jfc = udma_u_deactive_jfc,
+	.free_jfc = udma_u_free_jfc,
 	.create_jfs = udma_u_create_jfs,
 	.modify_jfs = udma_u_modify_jfs,
 	.query_jfs = udma_u_query_jfs,
 	.flush_jfs = udma_u_flush_jfs,
 	.delete_jfs = udma_u_delete_jfs,
 	.delete_jfs_batch = udma_u_delete_jfs_batch,
+	.alloc_jfs = udma_u_alloc_jfs,
+	.set_jfs_opt = udma_u_set_jfs_opt,
+	.active_jfs = udma_u_active_jfs,
+	.get_jfs_opt = udma_u_get_jfs_opt,
+	.deactive_jfs = udma_u_deactive_jfs,
+	.free_jfs = udma_u_free_jfs,
 	.create_jfr = udma_u_create_jfr,
 	.modify_jfr = udma_u_modify_jfr,
 	.query_jfr = udma_u_query_jfr,
 	.delete_jfr = udma_u_delete_jfr,
 	.delete_jfr_batch = udma_u_delete_jfr_batch,
 	.unimport_jfr = udma_u_unimport_jfr,
+	.alloc_jfr = udma_u_alloc_jfr,
+	.set_jfr_opt = udma_u_set_jfr_opt,
+	.active_jfr = udma_u_active_jfr,
+	.get_jfr_opt = udma_u_get_jfr_opt,
+	.deactive_jfr = udma_u_deactive_jfr,
+	.free_jfr = udma_u_free_jfr,
 	.create_jetty = udma_u_create_jetty,
 	.modify_jetty = udma_u_modify_jetty,
 	.query_jetty = udma_u_query_jetty,
@@ -48,6 +72,12 @@ static urma_ops_t g_udma_ops = {
 	.delete_jetty_batch = udma_u_delete_jetty_batch,
 	.unimport_jetty = udma_u_unimport_jetty,
 	.unbind_jetty = udma_u_unbind_jetty,
+	.alloc_jetty = udma_u_alloc_jetty,
+	.set_jetty_opt = udma_u_set_jetty_opt,
+	.active_jetty = udma_u_active_jetty,
+	.get_jetty_opt = udma_u_get_jetty_opt,
+	.deactive_jetty = udma_u_deactive_jetty,
+	.free_jetty = udma_u_free_jetty,
 	.create_jetty_grp = udma_u_create_jetty_grp,
 	.delete_jetty_grp = udma_u_delete_jetty_grp,
 	.create_jfce = udma_u_create_jfce,
@@ -116,10 +146,51 @@ static void udma_u_init_context(struct udma_u_context *udma_ctx,
 	udma_ctx->die_id = resp->die_id;
 	udma_ctx->dump_aux_info = resp->dump_aux_info;
 	udma_ctx->jfr_sge = resp->jfr_sge;
+	udma_ctx->sq_reserved = resp->sq_reserved;
 	udma_ctx->hugepage_enable = resp->hugepage_enable;
 	udma_ctx->sva_sep_mode_en = resp->sva_sep_mode_en;
 }
 
+static int udma_u_reserved_sq(struct udma_u_context *udma_ctx, struct udma_create_ctx_resp *resp)
+{
+	off_t offset = get_mmap_offset(0, udma_ctx->page_size, UDMA_MMAP_RESERVED_SQ);
+
+	(void)pthread_mutex_lock(&g_sq_reserved_lock);
+	if (g_sq_reserved_va) {
+		g_sq_reserved_refcount++;
+		(void)pthread_mutex_unlock(&g_sq_reserved_lock);
+		return 0;
+	}
+
+	// TODO: the mmap will be moved to the DTU
+	g_sq_reserved_va =
+		mmap((void *)(uintptr_t)resp->sq_reserved_va, resp->sq_reserved_len,
+			PROT_READ | PROT_WRITE, MAP_SHARED, udma_ctx->urma_ctx.dev_fd, offset);
+	if (g_sq_reserved_va == MAP_FAILED) {
+		g_sq_reserved_va = NULL;
+		(void)pthread_mutex_unlock(&g_sq_reserved_lock);
+		UDMA_LOG_ERR("failed to reserved sq, errno:%d, strerror:%s.\n",
+			     errno, strerror(errno));
+		return EINVAL;
+	}
+	g_sq_reserved_len = resp->sq_reserved_len;
+	g_sq_reserved_refcount++;
+	(void)pthread_mutex_unlock(&g_sq_reserved_lock);
+
+	return 0;
+}
+
+static void udma_u_free_reserved_sq(void)
+{
+	(void)pthread_mutex_lock(&g_sq_reserved_lock);
+	g_sq_reserved_refcount--;
+	if (g_sq_reserved_va != NULL && g_sq_reserved_refcount == 0) {
+		munmap(g_sq_reserved_va, g_sq_reserved_len);
+		g_sq_reserved_va = NULL;
+	}
+	(void)pthread_mutex_unlock(&g_sq_reserved_lock);
+}
+
 static void udma_u_destroy_jt_table(struct udma_u_context *udma_u_ctx)
 {
 	pthread_rwlock_destroy(&udma_u_ctx->jfr_table_lock);
@@ -185,10 +256,15 @@ static urma_context_t *udma_u_create_context(urma_device_t *dev, uint32_t eid_in
 		goto err_alloc_db;
 	}
 
+	if (udma_ctx->sq_reserved && udma_u_reserved_sq(udma_ctx, &resp))
+		goto err_reserved_sq;
+
 	udma_u_init_jetty_table(udma_ctx);
 
 	return &udma_ctx->urma_ctx;
 
+err_reserved_sq:
+	udma_u_free_db(&udma_ctx->urma_ctx, &udma_ctx->db);
 err_alloc_db:
 	(void)urma_cmd_delete_context(&udma_ctx->urma_ctx);
 err_create_ctx:
@@ -207,6 +283,7 @@ static urma_status_t udma_u_delete_context(urma_context_t *ctx)
 	urma_status_t ret = URMA_SUCCESS;
 
 	udma_u_destroy_jt_table(udma_ctx);
+	udma_u_free_reserved_sq();
 	udma_u_free_db(ctx, &udma_ctx->db);
 	udma_u_destroy_hugepage(udma_ctx);
 
-- 
2.28.0.windows.1

